_posts/2020-09-21-centralised-schema-management-microservices.md

```diff
@@ -0,0 +1,209 @@
+
+# A portrait of a CSM in three acts
+
+## The MVP
+
+To be of any use, a CSM implementation, whether homebrew or off-the-shelf, needs to include at least the following:
+
+- __Some way to distribute schemas__ in a format or formats that are understood by all parts of the system.
+   - Your options format-wise include [JSON Schema](https://json-schema.org/), [Apache Avro](https://avro.apache.org/docs/current/spec.html), [Protobuf](https://developers.google.com/protocol-buffers/docs/proto) schemas, and the comedy options of XML Schema or RELAX-NG. (Not going to link to those two.)
+   - Ideally you define your schema just once, and distribute it in multiple formats.
+   - If you have, or suspect you'll have in the future any external APIs, then your selection narrows. There are not many formats that are widely accepted, and JSON Schema is probably the least problematic option.
+- For each of your services, a means of __consuming (using) the schema__. Having multiple formats available makes it easier to find a good consumer or validator.
+   - It's a nice bonus if your consumer can process the schema into something native for your under-the-hood implementation. This includes F# type providers, Clojure spec or Plumatic schema, or even just plain old Java code generators. (Let's not mention JAXB.)
+   - If you can do this runtime, that's even better. But you can often get by with compile-time consumption.
+   - At a minimum, your consumer must be able to tell you __if a data structure is valid or not__ in the context of a schema.
+   - If it can also do __coercion__ of _almost_ valid data into _completely_ valid data, like parsing an ISO8601 string into a DateTime object, that's good.
+- Schema distribution to consumers can be done during compilation or packaging, but it's more microservicey if you have a __schema registry service__. It can just be nginx serving static files, in a pinch.
+
+## We're not in MVP land anymore
+
+The above is fine for a workshop demo. However, it's likely your system will need something a bit more advanced to survive in the wild. To upgrade your CSM from the "Home Edition" into "Professional", you'll probably need these:
+
+- __Support for schema evolution__. Schemas are like war plans, they never survive enemy contact. You need to prepare for this by having a way to handle changes to your schemas without the need to rebuild everything and breaking all your external APIs.
+  - But what to do if you end up with an object of version 1, and you need to upgrade it into version 3? Or the reverse?
+  - __Transform__ is what you do. Your CSM system can help by providing transformation services, or supplying transformation instructions alongside your schemas. To do this efficiently and in a way compatible with as many languages and platforms as possible is difficult, though. More on this later on.
```

Comment:
Hmm... Is this like a traditional Enterprise Service Bus providing transformation services to XML messages between different versions? I though that was an architectural anti-pattern, doomed to encode business-logic into the service bus, and doomed to produce messages full of weird placeholder structures to encode for information that simply isn't available? Maybe you should explain a bit why this is a good thing and what kind of problems it brings? Maybe explain a bit how it is _not_ like the feared and loathed ESB?